영상 상세 페이지

hyokualexkwon opened the pull request:
## 작업 내용

<!-- 작업 내용을 상세하게 기술한다. -->
<!-- what 보단 why 와 how에 대해 서술한다 -->

### 서버

- 한 개의 비디오 데이터를 반환하는 API를 작성했습니다.
- 비디오의 댓글 데이터를 반환하는 API를 작성했습니다.
- 한 개의 댓글에 작성된 답글 데이터를 반환하는 API를 작성했습니다.

### 클라이언트

- 비디오 상세 페이지에서 비디오 데이터, 댓글 데이터, 답글 데이터를 호출하고 화면에 그리는 로직을 작성했습니다.
- 비디오 데이터가 로딩중일 때, 데이터가 표시될 부분을 skeleton을 작성했습니다.
- 댓글 데이터 또는 답글 데이터를 로딩할 때, 원형 progress ui를 사용하도록 작성했습니다.
- 댓글의 페이지네이션에 infinite scroll 기능을 추가했습니다. 이 부분은 react-infinite-scroll-component 모듈을 사용했습니다.

## 실행 방법

<!-- 일반적인 실행 방법과 다른 경우 이에 대해 명시한다. -->

## 결과 첨부

<!-- 자신의 작업 결과물과 리뷰어가 실행한 작업 결과물이 다를 수 있으므로 본인의 결과물을 사진등으로 첨부한다. -->
![image](https://user-images.githubusercontent.com/17868599/69895479-af36ea80-1373-11ea-9d5c-c22effd6dc09.png)

hyokualexkwon committed:
[client] | refactor(Tabs): Filter 컴포넌트를 Tabs로 명명

기존에 사용하던 Filter 를 Tabs로 이름을 변경했습니다. 필터라기보다는 탭 ui를 그려내는 컴포넌트 이기 때문에 조금 더 일반적인 이름을 사용했습니다.

hyokualexkwon committed:
[server] | refactor(): dto class 수정

생성자의 인자로 들어오는 video에 private readonly 키워드를 적용했을 때에는 this에 바인딩 되므로 해당 키워드를 삭제했습니다.

hyokualexkwon committed:
[server] | refactor(video.controller): response 객체 변환

getVideos가 반환하는 배열의 각 항목을 VideoResponseDto 객체로 변환합니다.

hyokualexkwon committed:
[server] | feat(findVideo): 한 개의 비디오 정보 API 작성

hyokualexkwon committed:
[server] | refactor(video.service) 공통 로직 분리

- VIDEO_QUERY_SELECT_COLUMNS를 별도로 분리했습니다.
- offset을 계산하는 getOffset을 별도의 유틸로 분리했습니다.

hyokualexkwon committed:
[server] | feat(comment): 모듈 및 서비스 생성

hyokualexkwon committed:
[server] | feat(comments): 댓글 관련 로직 추가

comment service에서 댓글 데이터를 반환하는 로직을 작성했습니다.

- findComment 메소드를 통해 한 개의 댓글 항목을 반환할 수 있습니다.
- findCommentsByVideo 메소드를 통해 영상 항목에 작성된 댓글 항목 리스트를 반환할 수 있습니다.
- findReplies메소드를 통해 한 개의 댓글 항목에 잓어된 대댓글 항목 리스트를 반환할 수 있습니다.

hyokualexkwon committed:
[server] | feat(comments): video 컨트롤러에서 댓글 반환 로직 추가

video controller 에서, 댓글 항목을 반환하는 로직을 작성했습니다. 주로 comment service를 활용하게 됩니다.

hyokualexkwon committed:
[server] | refactor(): comments 컨트롤러 pipe 작성

comments에 사용되는 parameter와 query에 대한 검증 및 데이터 형 변환 pipe를 작성했습니다.

hyokualexkwon committed:
[server] | refactor(): replies 컨트롤러 pipe 생성

hyokualexkwon committed:
[server] | refactor(videos): 복수개의 영상 리소스 반환 값에 갯수 추가

hyokualexkwon committed:
[server] | refactor(comments): 복수개의 댓글 리소스에 반환 값에 갯수 추가

hyokualexkwon committed:
[client] | feat(material-ui): lab 모듈 설치

material-ui에서 제공하는 skeleton 을 사용하기 위해 모듈을 추가했습니다.

hyokualexkwon committed:
[client] | refactor(videos): 비디오 리스트 에러 예외처리

영상 리스트 페이지들에서 에러가 발생했을 때 발생하는 런타임 에러에 대한 예외처리를 추가했습니다.

hyokualexkwon committed:
[client] | feat(video): 영상 상세 페이지 영상 데이터 호출

영상 상세 페이지에서 접근한 페이지에 사용되는 데이터를 호출하는 로직을 hook으로 작성했습니다.

hyokualexkwon committed:
[client] | refactor(video): 영상 상세 데이터 렌더링

영상 상세 정보를 그리는 VideoInfo와 VideoMeta 컴포넌트에 데이터를 그리도록 작성했습니다. 또, 데이터가 준비되지 않은 상태에서 스켈레톤 UI를 그리도록 재작성했습니다.

hyokualexkwon committed:
[client] | feat(comments): 댓글 항목 데이터 렌더링

리모트 서버에서 데이터를 응답받고, 해당 데이터로 댓글 리스트를 그리는 로직과 뷰를 작성했습니다.

hyokualexkwon committed:
[clinet] | feat(svg): svg 추가

hyokualexkwon committed:
[client] | refactor(CommentItem): 댓글 항목 컴포넌트 수정

한 개의 댓글 항목과 관계된 답글 리스트를 필요에 따라 호출하는 로직을 작성하고, 마크업과 스타일을 수정했습니다.

hyokualexkwon committed:
[client] | refactor(CommentForm): 외부에서 제어하도록 수정

컴포넌트의 행동을 외부에서 제어할 수 있도록 수정했습니다.

hyokualexkwon committed:
[clinet] | refactor(Video): 댓글 입력 행동 제어 로직 작성

댓글 양식을 제어하는 hook을 작성하고, CommentForm 을 제어합니다.

hyokualexkwon committed:
[client] | refactor(CommentItem):  답글 입력 양식 추가

Atercatus commented on packages/server/src/video/video.controller.ts:
```diff
@@ -27,13 +38,7 @@ export class VideoController {
   public async getVideos(
     @Query() videosQueryDto: VideosQueryDto,
   ): Promise<VideoResponseDto[]> {
```
하나의 DTO로 만들어야 될 거 같습니다.

hyokualexkwon replied:
허허 만들어놓고 적용을 못했네요~

Atercatus commented on packages/server/src/comment/comment.service.ts:
```diff
@@ -1,4 +1,80 @@
+      return qb.orderBy('Comment_popularity', 'DESC').getMany();
+    }
+
+    return qb.orderBy('Comment_createdAt', 'DESC').getMany();
```
`Comment_createdAt` 는 어디서 나온거에요?

hyokualexkwon replied:
Comment table의 createAt이요? 본 쿼리는 최신순 정렬을 하고있습니다

Atercatus commented on packages/server/src/video/pipe/comments-param-pipe.ts:
```diff
@@ -0,0 +1,23 @@
+    return { id: parseInt(id, 10) };
+  }
+
+  private validateParam(id: string) {
```
어떤 validate인 지 알 수 있는 네이밍도 좋을 거 같습니다.

hyokualexkwon replied:
고민해보겠습니다~

Atercatus commented on packages/server/src/video/pipe/replies-param-pipe.ts:
```diff
@@ -0,0 +1,27 @@
+
+  private validateId(id: string) {
+    const parsedId = parseInt(id, 10);
+    return !isNaN(parsedId) && parsedId > 0;
```
`isDigit()` 어떤가여

hyokualexkwon replied:
그게 무슨 메소드에요? 찾아봐도 못찾곘네요..

Atercatus commented on packages/client/components/CommentItem/styles.tsx:
```diff
@@ -99,23 +146,45 @@ export const ShowRepliesButton = styled.div`
+  display: flex;
+  justify-content: flex-start;
+
+  > div {
```
이렇게 하시는 이유가 있나여

hyokualexkwon replied:
이름 따로 안줄라구요~

KimJye commented on packages/server/src/video/video.service.ts:
```diff
@@ -21,34 +23,23 @@ export class VideoService {
 
     if (sort === LATEST) {
-      return await qb.orderBy('Video_createdAt', 'DESC').getMany();
+      return await qb.orderBy('Video_createdAt', 'DESC').getManyAndCount();
```
count하는 이유는 무엇인가요??

hyokualexkwon replied:
화면에서 count를 사용하기 위함이죠!

riverandeye commented on packages/server/src/video/video.controller.ts:
```diff
@@ -26,15 +51,96 @@ export class VideoController {
+  }
+
+  @Get('/:id')
+  public async getVideo(@Param('id') id: string): Promise<VideoResponseDto> {
```
이 친구도 Pipe로 validate 해야할것 같아요 :smile:

hyokualexkwon replied:
추가했습니다!

riverandeye commented on packages/server/src/video/video.controller.ts:
```diff
@@ -26,15 +51,96 @@ export class VideoController {
+    return new VideoResponseDto(video);
+  }
+
+  @Get('/:id/comments')
+  public async getComments(
+    @Param(null, new CommentsParamPipe()) commentsParamDto: CommentsParamDto,
+    @Query(null, new CommentsQueryPipe()) commentsQueryDto: CommentsQueryDto,
+  ): Promise<CommentsResponseDto> {
```
이 친구도 Pipe로 validate 해야할것 같아요 😄
할게 왜케 많죠??

hyokualexkwon replied:
pipe로 validate 하고 있습니다.
```
@Param(null, new CommentsParamPipe()) commentsParamDto: CommentsParamDto,
@Query(null, new CommentsQueryPipe()) commentsQueryDto: CommentsQueryDto,
```

riverandeye commented on packages/server/src/video/video.controller.ts:
```diff
@@ -26,15 +51,96 @@ export class VideoController {
+    };
   }
+
+  @Get('/:id/comments/:commentId/replies')
+  public async getReplies(
+    @Param(null, new RepliesParamPipe()) repliesParamDto: RepliesParamDto,
+    @Query(null, new RepliesQueryPipe()) repliesQueryDto: RepliesQueryDto,
```
이 친구도 Pipe로 validate 해야할것 같아요 😄
할게 너무많네요.. 추후에 하는것도 좋을거같아요..

hyokualexkwon replied:
```
  @Param(null, new RepliesParamPipe()) repliesParamDto: RepliesParamDto,
    @Query(null, new RepliesQueryPipe()) repliesQueryDto: RepliesQueryDto,
```

riverandeye commented on packages/server/src/video/video.controller.ts:
```diff
@@ -1,17 +1,42 @@
+import { RepliesParamPipe } from './pipe/replies-param-pipe';
+import { RepliesParamDto } from './dto/replies-param.dto';
+import { RepliesQueryPipe } from './pipe/replies-query-pipe';
+import { RepliesQueryDto } from './dto/replies-query.dto';
```
미리 작성된 index.ts에서 추가해주시면 좋을거같아요~~ 😄

hyokualexkwon replied:
차라리 이런 컨벤션이 없으면 좋겠네요. 일일히 파일과 함수를 작성해야 할 때 마다 index.ts에 포함해야하는데, 의미 없어 보입니다.

hyokualexkwon replied:
아예 index.ts를 없애도록 하겠습니다.

riverandeye reviewed (commented):
혹시, VideoComments가 처음 렌더될 때 함수가 6번 re-render 되는데, 이 부분은 어떤 이유에서 그런건가요?? (궁금)

hyokualexkwon committed:
[server] | refactor(video dto): 모듈 export하는 index.ts 삭제

hyokualexkwon committed:
[server] | refactor(): videos query pipe 수정

common에 위치했던 pipe를 videos 모듈에 포함했습니다. 동시에 조건문 로직을 한 데 정리했습니다.

hyokualexkwon committed:
[server] | refactor(Video): pipe 적용

한 개의 비디오 항목 반환 컨트롤러의 param에 pipe를 적용했습니다.

hyokualexkwon commented:
> 혹시, VideoComments가 처음 렌더될 때 함수가 6번 re-render 되는데, 이 부분은 어떤 이유에서 그런건가요?? (궁금)

반복적인 리렌더링은 의도하지 않은 현상입니다. 차후에 성능 튜닝단계에서 잡아나가면 될 것 같습니다.

hyokualexkwon committed:
Merge branch 'develop' into feature/video_item_20191125
